### ISSUE TITLE
Make sure S3 upload completion is idempotent

### ISSUE THREAD

**Issue by xemul** [scylladb/scylladb#19533]
A theory: when issuing POST /object?uploadId=foo request server may (?) seal the object, but fail to return us the response with success status. If the failure renders into transport error (connection closed by server) , seastar will try replay the request again.

refs: #19461

**Comment by kreuzerkrieg** [comment]
I would even extend this issue to completion AND abortion. Since most likely the handling will be in the same place

**Comment by kreuzerkrieg** [comment]
Ok, we have a problem here.
According to the [documentation](https://docs.aws.amazon.com/AmazonS3/latest/API/API_CompleteMultipartUpload.html) you can get `200 OK` and the error is embedded into response body
Example

```
 HTTP/1.1 200 OK
 x-amz-id-2: Uuag1LuByRx9e6j5Onimru9pO4ZVKnJ2Qz7/C1NPcfTWAtRPfTaOFg==
 x-amz-request-id: 656c76696e6727732072657175657374
 Date:  Mon, 1 Nov 2010 20:34:56 GMT
 Connection: close
 Server: AmazonS3

 <?xml version="1.0" encoding="UTF-8"?>

 <Error>
  <Code>InternalError</Code>
  <Message>We encountered an internal error. Please try again.</Message>
  <RequestId>656c76696e6727732072657175657374</RequestId>
  <HostId>Uuag1LuByRx9e6j5Onimru9pO4ZVKnJ2Qz7/C1NPcfTWAtRPfTaOFg==</HostId>
 </Error>
```
[Here](https://github.com/scylladb/scylladb/blob/86bc5b11fe94d488a2d0ffb78fdc923ed9983dac/utils/s3/client.cc#L706) in `upload_sink_base::finalize_upload` we call `_client->make_request(std::move(req));` with default `reply_handler` which is `ignore_reply`
Meaning we will consider the multipart upload completed successfully even when it failed.
Will create a PR to address it
@xemul FYI

P.S. Aborting is completely different story, so I retract my previous statement about extending this issue to also handle the AbortMultipartUpload, will create another issu

**Comment by xemul** [comment]
> According to the [documentation](https://docs.aws.amazon.com/AmazonS3/latest/API/API_CompleteMultipartUpload.html) you can get 200 OK and the error is embedded into response body

:vomiting_face: that's another (though related) issue that should be fixed too

**Comment by kreuzerkrieg** [comment]
> > According to the [documentation](https://docs.aws.amazon.com/AmazonS3/latest/API/API_CompleteMultipartUpload.html) you can get 200 OK and the error is embedded into response body
> 
> 🤮 that's another (though related) issue that should be fixed too

Right, opening another issue

**Comment by kreuzerkrieg** [comment]
Ok, regarding this very issue. According the aforementioned documentation

> Error Code: NoSuchUpload
> 
> Description: The specified multipart upload does not exist. The upload ID might be invalid, or the multipart upload might have been aborted or **_completed_**.
> 
> HTTP Status Code: 404 Not Found

@xemul looks like we dont have to do anything particular in this case too.
Closing?

**Comment by xemul** [comment]
No. Imagine Scylla is flushing an sstable. It creates upload, puts some parts there, then completes it. Completing goes wrong, server succeeds, but client gets connection failure and retries the completion request. Server fails the request, the whole flushing fails, but an sstable exists on the server.

**Comment by regevran** [comment]
> No. Imagine Scylla is flushing an sstable. It creates upload, puts some parts there, then completes it. Completing goes wrong, server succeeds, but client gets connection failure and retries the completion request. Server fails the request, the whole flushing fails, but an sstable exists on the server.

This is for a future "flush to S3" correct?

**Comment by regevran** [comment]
Will be closed when #21054

**Comment by regevran** [comment]
Waiting for a test by @kreuzerkrieg.
